### ISSUE TITLE
Memory issue

### ISSUE THREAD

**Issue by ukcroupier** [dmsc/fastbasic#7]
I'm having a problem with FB and FBC crashing, although my program works fine when I use the cross compiler.  So I'd like to rule out something before looking into it further.

Is page6 free for the user?  I'm using it for a short assembler script but if FastBasic uses it then this could be the problem.

Also, how much memory does FastBasic leave for the user?  I've ran ?fre() and I get:

Using FB = 29983
Using FBI = 31263
Using FBC = -28897
Using cross compiler = 31263

Is there really less than 31k for code/data?  I thought they'd be more.  Although the signed interger is clearly confusing me after compiling with FBC.

**Comment by ukcroupier** [comment]
[CTEST.txt](https://github.com/dmsc/fastbasic/files/3597692/CTEST.txt)

I've had another look at this and here's what I found (I've attached the program in txt file)

The attached program is slightly altered to disable the machine code in page6 and simply display a counter to show that it's running, but the original program with the machine code compiles and runs fine with the cross compiler.

The altered code will not run in FB and it throws up errors when compiled with FBC.  However, if you activate the first 2 lines which declare variables it will compile and run with FBC.

The strange thing is that when the program was a little shorter it compiled fine without having to declare the variables at the start.

I guess I need to put my machine code somewhere else, but the issue about needing to declare variables depending upon the size of the code seems strange.

**Comment by dmsc** [comment]
Hi!

> I'm having a problem with FB and FBC crashing, although my program works fine when I use the cross compiler. So I'd like to rule out something before looking into it further.
> 
> Is page6 free for the user? I'm using it for a short assembler script but if FastBasic uses it then this could be the problem.

FastBasic has the same usage as Atari Basic, using addresses from $580 (half of page 5) as INPUT buffer. This means that lower part of page 6 can be overwritten if you INPUT more than 128 characters, or if you use an OPEN or XIO argument with more than 128 characters.

> Also, how much memory does FastBasic leave for the user? I've ran ?fre() and I get:
> 
> Using FB = 29983
> Using FBI = 31263
> Using FBC = -28897
> Using cross compiler = 31263
> 
> Is there really less than 31k for code/data? I thought they'd be more.

The Atari architecture allows a maximum of 48K RAM to applications, but the DOS uses the first 8K and the Graphics 0 screen uses 1K, so you have a maximum of 39K available to users.

Over this, using FBC adds the full run-time of about 3K, so you have about 36K left for your application.

> Using cross compiler = 31263

This is wrong, the cross compiler should have left you with about 38.5K (so, reported as about -26300). I suspect that you tested with Atari BASIC enabled, so you have 8K less!.

> Although the signed interger is clearly confusing me after compiling with FBC.

Yes, I know. Perhaps FastBasic could differentiate between integers (from -32768 to 32767) and "words" (from 0 to 65535), but this would complicate the parser and also the language.

**Comment by dmsc** [comment]
Hi again!

> [CTEST.txt](https://github.com/dmsc/fastbasic/files/3597692/CTEST.txt)
> 
> I've had another look at this and here's what I found (I've attached the program in txt file)
> 
> The attached program is slightly altered to disable the machine code in page6 and simply display a counter to show that it's running, but the original program with the machine code compiles and runs fine with the cross compiler.
> 
> The altered code will not run in FB and it throws up errors when compiled with FBC. However, if you activate the first 2 lines which declare variables it will compile and run with FBC.
> 
> The strange thing is that when the program was a little shorter it compiled fine without having to declare the variables at the start.
> 

So, you uncovered a bug in the compiler - it was not properly detecting out of memory conditions, so depending on the structure of the code sometimes the code was simply ignored instead of compiled.

This is fixed now in commits 74f8e95637cb1c06a848b28d8c209610118c7136 and c3d245a8f21f6e69f372db20713014bc29bb2101 , attached is the compiled ATR.

Sadly, your program is too big for the native IDE (FB.COM) and compiler (FBC.COM), as the data statements consume too much memory.

You can use the following techniques to make your program shorter:
- Converting the DATA statements from BYTE to WORD, as this would be a smaller line,
- Instead of copying the data with "FOR/NEXT", you can simply use "MOVE", this is faster and smaller.
- Instead of setting memory with a "FOR/NEXT", you can use "MSET".
- Use `DPEEK(X)` instead of `PEEK(X)+PEEK(X+1)*256`.
- Replace nested IF with combined conditions:
   ```
  	IF X>72
 	IF X<175
 	IF Y>104
 	IF Y<152
 	IF Y1
 		TURNOK=0
 	ENDIF
 	ENDIF
 	ENDIF
 	ENDIF
 	ENDIF
   ```
   with
   ```
  	IF X>72 AND X<175 AND Y>104 AND Y<152 AND Y1
 		TURNOK=0
 	ENDIF
   ```
See attached program that "just" compiles with `FBC`.

[fastbasic.zip](https://github.com/dmsc/fastbasic/files/3598521/fastbasic.zip)
[CTEST.txt](https://github.com/dmsc/fastbasic/files/3598522/CTEST.txt)

**Comment by ukcroupier** [comment]
Thanks for the quick fix and the code suggestions, although those crazy IF combinations (and some other ugly coding) are designed to get a little more speed (I originaly wrote them as you suggested).   Although they don't have as big an effect after being cross compiled, 

I actually have all those data statements as seperate files, so it was compiling fine with FBC.

And yes, I was running my cross compiled xex with basic installed in emulator,   It now shows the amount of RAM I was expecting.
